solr/core/src/java/org/apache/solr/handler/admin/api/RestoreCoreAPI.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.admin.api;
+
+import static org.apache.solr.client.solrj.SolrRequest.METHOD.POST;
+import static org.apache.solr.common.params.CoreAdminParams.ACTION;
+import static org.apache.solr.common.params.CoreAdminParams.CORE;
+import static org.apache.solr.common.params.CoreAdminParams.CoreAdminAction.RESTORECORE;
+import static org.apache.solr.handler.ClusterAPI.wrapParams;
+import static org.apache.solr.security.PermissionNameProvider.Name.CORE_EDIT_PERM;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import org.apache.solr.api.Command;
+import org.apache.solr.api.EndPoint;
+import org.apache.solr.api.PayloadObj;
+import org.apache.solr.client.solrj.request.beans.RestoreCorePayload;
+import org.apache.solr.handler.admin.CoreAdminHandler;
+
+/**
+ * Internal V2 API triggering restore a core.
+ *
+ * <p>Only valid in SolrCloud mode. This API (POST /api/cores/coreName/restore {'restore': {}}) is
+ * analogous to the v1 GET /solr/admin/cores?action=RESTORECORE command.
+ *
+ * @see RestoreCorePayload
+ */
+@EndPoint(
+    path = {"/cores/{core}"},
+    method = POST,
+    permission = CORE_EDIT_PERM)
+public class RestoreCoreAPI {
+
+  public static final String V2_CORE_RESTORE_CMD = "restore";
+
+  private final CoreAdminHandler coreAdminHandler;
+
+  public RestoreCoreAPI(CoreAdminHandler coreAdminHandler) {
+    this.coreAdminHandler = coreAdminHandler;
+  }
+
+  @Command(name = V2_CORE_RESTORE_CMD)
```

Comment:
[-1] Similar my comment above about the API path, I think the overall shape of this code looks great.  Just need to tweak a few specifics so that the v2 RESTORECORE API lines up better with the direction we're trying to move our v2 APIs in.

Particularly, we're trying to move away from the v2 "command-based" APIs (i.e. APIs with a request body whose only top-level key is the name of some command or other).  Instead we're generally moving the command itself into the path (see my comment above about changing the path to `/cores/coreName/restore`) and then moving most of the actual restore parameters to the top level of the "Payload"/request-body.

This is definitely do-able using the Solr-custom annotation framework you've used in this PR (see RenameCollectionAPI for an example).  But you might find it more natural to do if you switch this API over to use the JAX-RS framework we've been moving towards.  If you want to try that out there's a step-by-step writeup on this [here](https://issues.apache.org/jira/browse/SOLR-15737?focusedCommentId=17617923&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17617923) and a handful of examples that you can crib from ([example1](https://github.com/apache/solr/pull/1452), [example2](https://github.com/apache/solr/pull/1441))